src/http/httpClient/Middleware/BodyInspectionHandler.cs

```diff
@@ -0,0 +1,95 @@
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Kiota.Http.HttpClientLibrary.Extensions;
+using Microsoft.Kiota.Http.HttpClientLibrary.Middleware.Options;
+
+namespace Microsoft.Kiota.Http.HttpClientLibrary.Middleware;
+
+/// <summary>
+/// The Body Inspection Handler allows the developer to inspect the body of the request and response.
+/// </summary>
+public class BodyInspectionHandler : DelegatingHandler
+{
+    private readonly BodyInspectionHandlerOption _defaultOptions;
+
+    /// <summary>
+    /// Create a new instance of <see cref="BodyInspectionHandler"/>
+    /// </summary>
+    /// <param name="defaultOptions">Default options to apply to the handler</param>
+    public BodyInspectionHandler(BodyInspectionHandlerOption? defaultOptions = null)
+    {
+        _defaultOptions = defaultOptions ?? new BodyInspectionHandlerOption();
+    }
+
+    /// <inheritdoc/>
+    protected override async Task<HttpResponseMessage> SendAsync(
+        HttpRequestMessage request,
+        CancellationToken cancellationToken
+    )
+    {
+        if (request == null)
+            throw new ArgumentNullException(nameof(request));
+
+        var options = request.GetRequestOption<BodyInspectionHandlerOption>() ?? _defaultOptions;
+
+        Activity? activity;
+        if (request.GetRequestOption<ObservabilityOptions>() is { } obsOptions)
+        {
+            var activitySource = ActivitySourceRegistry.DefaultInstance.GetOrCreateActivitySource(
+                obsOptions.TracerInstrumentationName
+            );
+            activity = activitySource?.StartActivity(
+                $"{nameof(BodyInspectionHandler)}_{nameof(SendAsync)}"
+            );
+            activity?.SetTag("com.microsoft.kiota.handler.bodyInspection.enable", true);
+        }
+        else
+        {
+            activity = null;
+        }
+        try
+        {
+            if (options.InspectRequestBody)
+            {
+                options.RequestBody = await CopyToStreamAsync(request.Content);
+            }
+            var response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+            if (options.InspectResponseBody)
+            {
+                options.ResponseBody = await CopyToStreamAsync(response.Content);
+            }
+
+            return response;
+        }
+        finally
+        {
+            activity?.Dispose();
+        }
+
+#if NETSTANDARD2_1_OR_GREATER || NETCOREAPP3_0_OR_GREATER
+        [return: NotNullIfNotNull(nameof(httpContent))]
+#endif
+        static async Task<Stream?> CopyToStreamAsync(HttpContent? httpContent)
+        {
+            if (httpContent == null)
+            {
+                return null;
+            }
+
+            var stream = new MemoryStream();
```

Comment:
I suspect there are scenarios where the response/request stream may not be seekable(not buffered) so setting the position or reading the stream twice will throw. We should probably add code to handle that (skip the copy)

Reply:
I modified the `CopyToStreamAsync` method, have a look

Reply:
Thanks. This makes sense. 
Do you think it would make sense to use this `TryComputeLength` instead?
https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpcontent.trycomputelength?view=net-8.0

Reply:
Sorry, I missed that. 
I believe we can go forward with the current one for now.